### PR TITLE
[ntuple] Implement reading and writing of sealed pages

### PR DESCRIPTION
**Pull request by jblomer**
One of the building blocks for fast merged, as sketched in #6105

### CODE REVIEW

File: tree/ntuple/v7/src/RPageStorageFile.cxx [hunk 1]
```diff
@@ -134,6 +134,28 @@ ROOT::Experimental::Detail::RPageSinkFile::CommitPageImpl(ColumnHandle_t columnH
...
+ROOT::Experimental::Detail::RPageSinkFile::CommitSealedPageImpl(
+   DescriptorId_t columnId, const RPageStorage::RSealedPage &sealedPage)
+{
+   const auto bitsOnStorage = Detail::RColumnElementBase::Generate(
+      fDescriptorBuilder.GetDescriptor()
+      .GetColumnDescriptor(columnId)
+      .GetModel()
+      .GetType()).GetBitsOnStorage();
+
+   const auto bytesPacked = (bitsOnStorage * sealedPage.fNElements + 7) / 8;
```
Comment (pcanal) [inline tree/ntuple/v7/src/RPageStorageFile.cxx:147]: Doesn't the sealed page already know this information (in some in the bytes header) and/or didn't we already calculated it when reading/loading the page?  [Recalculating it wouldn't even be note worthy it was not for th fact that it seems to require an allocation to get the base information).

File: tree/ntuple/v7/src/RPageStorageFile.cxx [hunk 2]
```diff
@@ -261,6 +283,42 @@ ROOT::Experimental::RNTupleDescriptor ROOT::Experimental::Detail::RPageSourceFil
...
+         pageInfo = pi;
+         break;
+      }
+      firstInPage += pi.fNElements;
+      ++pageNo;
+   }
+   R__ASSERT(firstInPage <= index);
+   R__ASSERT((firstInPage + pageInfo.fNElements) > index);
+
+   const auto bytesOnStorage = pageInfo.fLocator.fBytesOnStorage;
```
Comment (pcanal) [inline tree/ntuple/v7/src/RPageStorageFile.cxx:311]: We should probably include this information in the RSealedPage struct.

Comment (pcanal) [reply tree/ntuple/v7/src/RPageStorageFile.cxx:311]: humm .. it's there already :).

File: tree/ntuple/v7/src/RPageStorageFile.cxx (hunk 1 above)
Comment (pcanal) [reply tree/ntuple/v7/src/RPageStorageFile.cxx:147]: reading more, isn't ```bitsOnStorage == sealedPage.fSize```?  humm .... I seems to be confused the variable name "bitsOnStorage" seems to be used to mean sometimes `number of bits of an individual elements` or `total size of the page on disk` ... humm ... I see `bits` vs `bytes` ... the factor fNElements between the 2 is counter-intuitive.

File: tree/ntuple/v7/test/ntuple_merger.cxx [hunk 3]
```diff
@@ -1,5 +1,70 @@
 #include "ntuple_test.hxx"
 
+namespace {
+
+// Reads an integer from a little-endian 4 byte buffer
+std::int32_t ReadRawInt(void *ptr)
```
Comment (pcanal) [inline tree/ntuple/v7/test/ntuple_merger.cxx:6]: Don't we already have this in a more generic place?

File: tree/ntuple/v7/test/ntuple_merger.cxx [hunk 4]
```diff
@@ -1,5 +1,70 @@
...
+} // anonymous namespace
+
+TEST(RPageStorage, ReadSealedPages)
+{
+   FileRaii fileGuard("test_ntuple_sealed_pages.root");
+
+   // Create an ntuple at least 2 clusters, one with 1 entry and one with 100000 entries.
+   // Hence the second cluster should have more than a single page per column.  We write uncompressed
+   // pages so that we can meaningfully peek into the content of read sealed pages later on.
+   auto modelReferende = RNTupleModel::Create();
```
Comment (pcanal) [inline tree/ntuple/v7/test/ntuple_merger.cxx:27]: Is `modelReferende` spelt as intended?

File: tree/ntuple/v7/test/ntuple_merger.cxx [hunk 5]
```diff
@@ -1,5 +1,70 @@
...
+   source.Attach();
+   const auto &descriptor = source.GetDescriptor();
+   auto columnId = descriptor.FindColumnId(descriptor.FindFieldId("pt"), 0);
+   RClusterIndex index(descriptor.FindClusterId(columnId, 0), 0);
+   auto sealedPage = source.ReadSealedPage(columnId, index);
+   ASSERT_EQ(1U, sealedPage.fNElements);
+   ASSERT_EQ(4U, sealedPage.fSize);
+   EXPECT_EQ(42, ReadRawInt(sealedPage.fBuffer.get()));
+
+   // Check second, big cluster
```
Comment (pcanal) [inline tree/ntuple/v7/test/ntuple_merger.cxx:52]: I can't find the "Check first" comment ... :)

File: tree/ntuple/v7/test/ntuple_merger.cxx (hunk 3 above)
Comment (pcanal) [reply tree/ntuple/v7/test/ntuple_merger.cxx:6]: > We have it as internal functions in RMiniFile.

What is the advantage of duplicating it in both place?

Comment (jblomer) [reply tree/ntuple/v7/test/ntuple_merger.cxx:6]: Reading a 4 byte little endian buffer seemed like a small and innocent function.  In `RMiniFile.cxx` we have the four classes `RUInt16BE`, `RUInt32BE`, `RInt32BE`, and `RUInt64BE` in an anonymous namespace that implement the endianess swap but slightly different, ensuring big endian representation.

Those classes make sense in the context of the mini file implementation.  If we take them out and make them available as general infrastructure classes, more work is needed, e.g. support for the `LE` counterparts.

**Review by pcanal — APPROVED** [review]

**Pull request merged**